x/subscription/keeper/cu_tracker.go

```diff
@@ -5,34 +5,179 @@ import (
 	cuTrackerKey := types.CuTrackerKey(sub, provider, chainID)
 	var trackedCu types.TrackedCu
-	found = k.cuTrackerFS.FindEntry(ctx, cuTrackerKey, block, &trackedCu)
-	return trackedCu.GetCu(), found, cuTrackerKey
+	entryBlock, found = k.cuTrackerFS.FindEntry2(ctx, cuTrackerKey, uint64(ctx.BlockHeight()), &trackedCu)
+	if !found {
+		// entry not found -> this is the first, so not an error. return the current block and CUs=0
+		return 0, uint64(ctx.BlockHeight()), found, cuTrackerKey
+	}
+	return trackedCu.Cu, entryBlock, found, cuTrackerKey
 }
 
-func (k Keeper) AddTrackedCu(ctx sdk.Context, sub string, provider string, chainID string, cu uint64, block uint64) error {
-	if sub == "" || provider == "" || block > uint64(ctx.BlockHeight()) {
-		return utils.LavaFormatError("cannot add tracked CU",
-			fmt.Errorf("sub/provider cannot be empty. block cannot be larger than current block"),
-			utils.Attribute{Key: "sub", Value: sub},
-			utils.Attribute{Key: "provider", Value: provider},
-			utils.Attribute{Key: "block", Value: strconv.FormatUint(block, 10)},
-		)
-	}
-	trackedCu, _, key := k.GetTrackedCu(ctx, sub, provider, chainID, block)
-	err := k.cuTrackerFS.AppendEntry(ctx, key, block, &types.TrackedCu{Cu: trackedCu + cu})
+// AddTrackedCu adds CU to the CU counters in relevant trackedCu entry
+// Note that the trackedCu entry always has one version (updating it using append in the same block acts as ModifyEntry)
+func (k Keeper) AddTrackedCu(ctx sdk.Context, sub string, provider string, chainID string, cuToAdd uint64) error {
+	cu, entryBlock, _, key := k.GetTrackedCu(ctx, sub, provider, chainID)
+	err := k.cuTrackerFS.AppendEntry(ctx, key, entryBlock, &types.TrackedCu{Cu: cu + cuToAdd})
 	if err != nil {
 		return utils.LavaFormatError("cannot add tracked CU", err,
 			utils.Attribute{Key: "tracked_cu_key", Value: key},
-			utils.Attribute{Key: "block", Value: strconv.FormatUint(block, 10)},
-			utils.Attribute{Key: "current_cu", Value: trackedCu},
-			utils.Attribute{Key: "cu_to_be_added", Value: strconv.FormatUint(cu, 10)})
+			utils.Attribute{Key: "sub_block", Value: strconv.FormatUint(entryBlock, 10)},
+			utils.Attribute{Key: "current_cu", Value: strconv.FormatUint(cu, 10)},
+			utils.Attribute{Key: "cu_to_be_added", Value: strconv.FormatUint(cuToAdd, 10)})
 	}
 	return nil
 }
+
+// GetAllSubTrackedCuIndices gets all the trackedCu entries that are related to a specific subscription
+func (k Keeper) GetAllSubTrackedCuIndices(ctx sdk.Context, sub string) []string {
+	return k.cuTrackerFS.GetAllEntryIndicesWithPrefix(ctx, sub)
+}
+
+// removeCuTracker removes a trackedCu entry
+func (k Keeper) resetCuTracker(ctx sdk.Context, sub string, info trackedCuInfo, shouldRemove bool) error {
+	key := types.CuTrackerKey(sub, info.provider, info.chainID)
+	if shouldRemove {
+		return k.cuTrackerFS.DelEntry(ctx, key, uint64(ctx.BlockHeight()))
```

Comment:
no it shouldn't. you can't delete a past entry. Since the "old" entry's block is the subscription block, DelEntry will fail. The current height is correct because it deletes the nearest-smaller entry, which is exactly the one we wish to delete (I verified all of this using unit tests too, not just by looking at code flow)

Reply:
isnt this a bug than?
since this function is called with a timer after a new month already began wouldn't've it delete the new entry?

Reply:
this is not a bug. added a deletion unit test to make sure there are no problems with deletions: https://github.com/lavanet/lava/commit/239655e854e6f96daad9dd123be82f99be9a7fab